lib/ansible/modules/system/ufw.py

```diff
@@ -303,13 +319,47 @@ def ufw_version():
         if command == 'state':
             states = {'enabled': 'enable', 'disabled': 'disable',
                       'reloaded': 'reload', 'reset': 'reset'}
-            execute(cmd + [['-f'], [states[value]]])
+            if module.check_mode:
+                if value in ['reloaded', 'reset']:
+                    changed = True
+                else:
+                    ufw_enabled = pre_state.find("active") != -1
+                    if (value == 'disabled' and ufw_enabled) or (value == 'enabled' and not ufw_enabled):
+                        changed = True
+            else:
+                execute(cmd + [['-f'], [states[value]]])
 
         elif command == 'logging':
-            execute(cmd + [[command], [value]])
+            if module.check_mode:
+                extract = re.search(r'Logging: (on|off) \(([a-z]+)\)', pre_state)
+                if extract:
+                    current_level = extract.group(2)
+                    current_on_off_value = extract.group(1)
+                    if value != "off":
+                        if value != "on" and value != current_level:
+                            changed = True
+                    elif current_on_off_value != "off":
+                        changed = True
+                else:
+                    changed = True
+
+            else:
+                execute(cmd + [[command], [value]])
 
         elif command == 'default':
-            execute(cmd + [[command], [value], [params['direction']]])
+            if module.check_mode:
+                extract = re.search(r'Default: (deny|allow) \(incoming\), (deny|allow) \(outgoing\), (deny|allow) \(routed\)', pre_state)
+                if extract:
+                    current_default_values = {}
+                    current_default_values["incoming"] = extract.group(1)
+                    current_default_values["outgoing"] = extract.group(2)
+                    current_default_values["routed"] = extract.group(3)
+                    if current_default_values[params['direction']] != value:
+                        changed = True
+                else:
+                    changed = True
+            else:
+                execute(cmd + [[command], [value], [params['direction']]])
```

Comment:
Same problem with check mode here.

Reply:
Actually, no, this one works.

(I did a mistake while testing; I'm currently adding more tests, you'll see the results soon. They will also contain tests for check mode, even though right now many of them are commented out because they give the wrong result...)

Reply:
I so your tests with docker, is it easy to execute ?

Reply:
Unfortunately `ufw` doesn't seem to work in docker containers, so you'd have to run it on your own machine, or in some VM.

I've put the following playbook (`run-tests-locally.yml`) into `test/integration/targets/ufw/` in my Ansible dev tree; executing it with `-b -K` (so it asks for `sudo` password) runs the test on localhost:
```.yaml
---
# Run with -b -K arguments to ansible-playbook!
- hosts: localhost
  tasks:
  # Run the tests
  - block:
    - include_tasks: tasks/run-test.yml
      with_fileglob:
      - "tasks/tests/*.yml"
    vars:
      role_path: '.'
    always:
    - name: Reset ufw to factory defaults
      ufw:
        state: reset
    - name: Enable ufw (with factory defaults)
      ufw:
        state: enabled
```
(You can also use `ansible-test` to run the tests, but I prefer this lightweight playbook, in particular since the docker feature of `ansible-test` is useless for this test since it doesn't work with docker containers...)

I don't have an important firewall config on my machine, so after `reset` and `enabled` it's back to the state I need. Depending on your firewall config and your network / services running on your machine, you don't want to run this on your local machine... :)

Reply:
Did you manage to cancel the locale ?

Reply:
In my future push, I added to your shell call :
```
environment:
    LC_ALL: C
```

Reply:
I'm using `en_US.UTF-8` as my default locale, so I usually don't have problems :) (I gave up using localized OS versions a long time ago, the strange and sometimes half-missing translations kept annoying me too much ;) )